### PR TITLE
Use Ubuntu 18.04

### LINKED ISSUE #3664: Require at least Ubuntu 18.04
OpenTitan currently uses Ubuntu 16.04 as oldest supported version of Ubuntu, but also as baseline for other distributions ("if they aren't older than Ubuntu 16.04, things should be fine"). This issue suggests moving the baseline to Ubuntu 18.04 LTS. 

Motivation for going from Ubuntu 16.04 to Ubuntu 18.04

* Ubuntu 16.04 (as the name implies) has been released in April 2016 and will reach its end of life in April 2021 (https://help.ubuntu.com/community/EOL#Ubuntu_16.04_Xenial_Xerus; extended support releases might be available after that)
* Primary motivation: Ubuntu 16.04 comes with Python 3.5, which "will" be [EOL in September 2020](https://www.python.org/downloads/release/python-3510/). Even though the Python package in Ubuntu will still receive security fixes, more and more third-party Python packages are dropping support for Python 3.5.
  * [Meson 0.57 will require Python 3.6](https://github.com/mesonbuild/meson/blob/master/docs/markdown/snippets/python35-eol.md)
  * pip (the python package installer) will drop Python 3.5-suppport soon ("DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.")
  * Python 3.6 provides some features which many developers have already been trying to use, and we had to go back and remove them: Pathlib in Python 3.6 is significantly easier to use, since Path objects are now accepted in calls like `open()`. f strings are supported. etc.
* Clang(-format) will be available in version 6.0 (vs. 3.8 in Ubuntu 16.04)
* (Host) GCC moves from 5.4 to 7.4.0

Discussion: I expect the transition to be smooth and mostly unnoticed by users.

* As so often, EDA tool support is a major reason to stay with older OS releases. However, Ubuntu 18.04 is already "old enough" so that existing workflows are expected to continue to work.
  * Vivado 2018.3+ supports Ubuntu 18.04. That means we do have a couple of old Vivado versions available if necessary.
  * Synopsys products in the P (2019) and Q (2020) platform releases claim to support "Debian 9.0" (released mid-2017) and have never officially supported Ubuntu. However, we know that VCS works well with Ubuntu 18.04, DC also typically works well. 
  * Cadence tools and more specialized design tools are often more tricky to get working under Ubuntu in the first place, and many of them require RHEL/CentOS. The Ubuntu version change shouldn't affect those users, as long as they are able to obtain a sufficiently recent Python version (e.g. from a RHEL SCL). This is no different from today.
  * None of the open source EDA tools we're using have any problem with newer Ubuntu versions.
* Many users (both on the Google side, and here at lowRISC) have been using a variety of newer distributions, including Ubuntu 18.04, for work for a long time now and have not seen issues.

How to do it:
1. Form consensus around this move.
2. Announce it publicly if agreed.
3. Update the documentation, and switch all CI base images to Ubuntu 18.04

Please comment if you have seen problems or success stories with Ubuntu 18.04, or have seen issues where components of Ubuntu 18.04 are too new for a certain tool to work.

### PR DESCRIPTION
**Pull request by imphil**
Require at least Ubuntu 18.04, or another distribution which provides components which are similar to the ones in Ubuntu 18.04.

Fixes #3664

### CODE REVIEW

**Comment by lenary** [comment]
This will need a rebase, but otherwise I'm feeling pretty positive about the change. 

I tried out 18.04 using the utilities docker container, and it doesn't look as if it will cause many issues if we enable the extra warnings in the software build that I'm working on, which seems good.

**Comment by imphil** [comment]
This PR is now rebased to include the azure-pipelines change that Sam did recently (the slow lint job). 

We do now have project-wide agreement in #3664 that this change should go ahead, hence this PR is ready for review. Can I get please a positive review from @gkelly @lenary @asb and @sjgitty?

CC @rbohn @rasmus-madsen @martin-lueker FYI

Once it merges I'll send an announcement out to the project's mailing list to also inform other OpenTitan users of this change in requirements.

**Comment by imphil** [comment]
I updated this PR to rebase it on top of #3757, which clarifies our system requirements. As result, we now more clearly see what changes:

Two packages have changed their minimum version requirement:
* Ninja 1.5.1 -> 1.8.2 
* Python 3.5.2 -> 3.6 [1]

In #3664 all OpenTitan core contributors indicated that these requirements are already met in their setup.

@lenary @gkelly This update leaves the C/C++ host compiler requirements untouched. Do you want to update these requirements as well here, e.g. to make use of newer GCC/Clang features, or is this something we can do later? This probably would move us away from requiring "any compiler which can do C++14" to "at least GCC version X and clang version Y", which is "less nice", but in practice more realistic and something I'd prefer we do. We have heard from @rasmus-madsen that they have GCC 9.1 available, so that shouldn't be a blocker.

@rasmus-madsen @martin-lueker please have a look.


[1] I did not specify an exact micro-version of Python 3.6, since we haven't figured out an exact requirement yet. We know of some contributors having Python 3.6.4 available, while Ubuntu 18.04 ships with Python 3.6.7 in the latest update. Since Python is generally rather good at not breaking compatibility between micro-releases, I'm proposing to give it a try and react if needed.

**Comment by lenary** [comment]
Let's leave the C/C++ host compiler requirements for the moment - the host software has not been a focus recently and so there haven't been lots of changes there. If this changes, we can update this.

File: doc/ug/install_instructions/index.md
Comment (lenary) [reply doc/ug/install_instructions/index.md:76]: There's a year of changes between 1.8.11 (the 16.04 version) and 1.8.13 (the 18.04 version). These include both bugfixes and changes according to the changelog. I'm happy as long as the docs still render.

**Review by rasmus-madsen — APPROVED** [review]
@imphil 
as you know we don't run Ubuntu, but we have most of the tool versions required.
so I don't foresee any new challenges with this requirement

**Review by lenary — APPROVED** [review]
LGTM!

**Review by asb — APPROVED** [review]
This looks good to me and I think we got feedback from everyone necessary in the companion issue thread. Thanks Philipp for driving this forwards.

**Review by gkelly — APPROVED** [review]
Thanks for getting this done @imphil!

**Review by rbohn — APPROVED** [review]

**Review by martin-lueker — APPROVED** [review]

**Comment by imphil** [comment]
Thanks for all your input! I'm merging this now, and will announce it on the mailing list shortly.

**Comment by lenary** [comment]
@gkelly can you push the new docs builder today?

**Pull request merged**
